index.js

```diff
@@ -100,185 +106,206 @@ const VerifyOptions = t.interface({
+    Prolong(prolong);
+
+    const accessExp = computeExpiryDate(prolong ? prolongedAccessTTL : regularAccessTTL);
+    const refreshTtl = prolong ? prolongedRefreshTTL : regularRefreshTTL;
+    // Order of spreading is important!
+    const payload = Payload({
+      ...defaultSignOptions,
+      ...SignOptions(signOptions),
+      pld: content,
+      uid: userId,
+      exp: accessExp,
+      rme: prolong,
+      jti: await getEncodedRandomBytes()
+    });
+
+    const refreshToken = await generateRefreshToken(userId);
+    const accessToken = jwt.sign(payload, Secret(secret), {algorithm});
+
+    await store.registerTokens(userId, refreshToken, accessToken, refreshTtl * 1000);
+
+    return {
+      accessToken,
+      accessTokenExpiresAt: accessExp,
+      refreshToken,
+      refreshTokenExpiresAt: computeExpiryDate(refreshTtl)
+    };
+
+  };
 
   /**
    * Verifies token, might throw jwt.verify errors
    * @param {String} token
-   * @param secret
+   * @param {Secret} secret
    * @param {VerifyOptions} [verifyOptions] Options to pass to jwt.verify.
    * @returns {Promise<String>} decoded token
    * @throws JsonWebTokenError
    * @throws TokenExpiredError
    * Error info at {@link https://www.npmjs.com/package/jsonwebtoken#errors--codes}
    */
-  verify(token, secret, verifyOptions = {}) {
-    return this._jwt.verify(Token(token), Secret(secret),
-      mergeAll([this._defaultVerifyOptions, VerifyOptions(verifyOptions),
-        {algorithm: this._algorithm}]));
-  }
+  const verify = (token, secret, verifyOptions = {}) =>
+    jwt.verify(Token(token), Secret(secret), {
+      ...defaultVerifyOptions,
+      ...VerifyOptions(verifyOptions),
+      algorithm: algorithm
+    });
 
   /**
    * Issues a new access token using a refresh token and an old token.
    * There is no need to verify the old token provided because this method uses the stored one.
    * @param {String} refreshToken
-   * @param {String} oldToken
-   * @param secret
-   * @param {UserSignOptions} [signOptions] Options passed to jwt.sign
+   * @param {String} accessToken
+   * @param {Secret} secret
+   * @param {SignOptions} [signOptions] Options passed to jwt.sign
    * @returns {Promise<Tokens>}
-   * @throws {RefreshTokenExpiredOrNotFound} RefreshTokenExpiredOrNotFound
-   * @throws {InvalidAccessToken} InvalidAccessToken
+   * @throws {RefreshTokenExpiredOrNotFound} refreshTokenExpiredOrNotFound
+   * @throws {InvalidAccessToken} invalidAccessToken
    */
-  async refresh(refreshToken, oldToken, secret, signOptions) {
+  const refresh = async (refreshToken, accessToken, secret) => {
     RefreshToken(refreshToken);
-    Token(oldToken);
-    const untrustedPayload = Payload(this._jwt.decode(oldToken));
-    const trustedToken =
-      await this._store.getAccessToken(untrustedPayload.pld.userId, refreshToken);
+    Token(accessToken);
+
+    const userId = getUserId(refreshToken);
+    const storedToken = await store.getAccessToken(userId, refreshToken);
```

Comment:
What happens if the store cannot find it?

Reply:
as discussed offline